lucene/core/src/java/org/apache/lucene/codecs/lucene86/Lucene86PointsWriter.java

```diff
@@ -68,11 +71,9 @@ public Lucene86PointsWriter(SegmentWriteState writeState, int maxPointsInLeafNod
                                  writeState.segmentInfo.getId(),
                                  writeState.segmentSuffix);
 
-      String metaFileName = IndexFileNames.segmentFileName(writeState.segmentInfo.name,
-          writeState.segmentSuffix,
-          Lucene86PointsFormat.META_EXTENSION);
-      metaOut = writeState.directory.createOutput(metaFileName, writeState.context);
-      CodecUtil.writeIndexHeader(metaOut,
+      tempMetaOut = writeState.directory.createTempOutput(
```

Comment:
Why are we switching to a temp file and copying to the real file after closing?  Maybe add a comment explaining?

Reply:
This is because we need to write file lengths of the index/data files before any offsets/lengths of slices into these files. But since these index/data files have not been written yet, we don't know the length yet. So I wrote into a temp file, and only then write the final metadata file that includes first the lengths of the index/data files and then metadata about the KD trees that includes offsets into these index/data files. I'll add a comment.

Reply:
As an alternative, I could buffer the metadata in memory like we do for terms. It will require changing some APIs to replace IndexOutput with DataOutputs but other than that it shouldn't be too hard.

Reply:
OK thanks for the explanation @jpountz.  I am OK with using temp files for this ...